cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,869 @@
+ * @tparam V the type of the right operand.
+ * @param x The left operand.
+ * @param y The right operand.
+ *
+ * @returns Pair of `x - y`, saturated at the numeric limits for the type of
+ * `x`, without overflowing or invoking undefined behaviour, and
+ * whether overflow occurred.
+ *
+ * @note If `T` is a numeric type we must have `std::is_same_v<T,
+ * V>`. If `T` is a timestamp type, `V` must be a duration type and
+ * `std::is_same_v<typename T::duration, V>`. Note in particular that the
+ * usual integral promotion rules are not applied. If `T` is a fixed
+ * point type, then `V` must be the representation type of `T`, and it
+ * is required that `x` and `y` have the same scale. If `T` is a
+ * floating point type, then it is required (not checked) that `y` is
+ * not inf or nan, otherwise behaviour is undefined, if `x` is finite,
+ * then overflow to +-inf is clamped at lowest()/max().
+ */
+template <typename T, typename V>
+[[nodiscard]] __host__ __device__ constexpr inline cuda::std::pair<T, bool> saturating_sub(
+  T x, V y) noexcept
+{
+  if constexpr (cudf::is_timestamp_t<T>()) {
+    static_assert(cudf::is_duration_t<V>(), "Can only add durations to timestamps");
+    static_assert(cuda::std::is_same_v<typename T::duration, V>,
+                  "Duration resolution must match timestamp resolution");
+    auto const [value, did_overflow] = saturating_sub(x.time_since_epoch(), y);
+    return {T{value}, did_overflow};
+  } else if constexpr (cudf::is_duration_t<T>()) {
+    static_assert(cuda::std::is_same_v<T, V>, "Cannot add mismatching types");
+    auto const [value, did_overflow] = saturating_sub(x.count(), y.count());
+    return {T{value}, did_overflow};
+  } else if constexpr (cudf::is_fixed_point<T>()) {
+    using Rep = typename T::rep;
+    // Requirement, not checked, x and y have the same scale.
+    static_assert(cuda::std::is_same_v<Rep, V>, "Must add rep type of fixed point to fixed point.");
+    auto const [value, did_overflow] = saturating_sub(x.value(), y);
+    return {T{numeric::scaled_integer<Rep>{value, x.scale()}}, did_overflow};
+  } else {
+    static_assert(cuda::std::is_same_v<T, V>, "Cannot add mismatching types");
+    if constexpr (cuda::std::is_floating_point_v<T>) {
+      // Mimicking spark requirements, inf/nan x propagates
+      if (cuda::std::isinf(x) || cuda::std::isnan(x)) { return {x, false}; }
+      // Requirement, not checked, y is not inf or nan.
+      T result = x - y;
+      // If the result is outside the range of finite values it can at
+      // this point only be +- infinity (we can't generate a nan by
+      // subtracting a non-nan/non-inf y from a non-nan/non-inf x).
+      if (result < cuda::std::numeric_limits<T>::lowest()) {
+        return {cuda::std::numeric_limits<T>::lowest(), true};
+      } else if (result > cuda::std::numeric_limits<T>::max()) {
+        return {cuda::std::numeric_limits<T>::max(), true};
+      }
+      return {result, false};
```

Comment:
@mythrocks Is there anywhere that the semantics of rolling window bounds in the presence of inf/nan in the orderby column are written down for spark?

Reply:
No, not window-function behaviour exactly.  I've had to derive some of these from experimenting with Spark. 

The arithmetic/ordering rules for special floating point values are [listed in the language manual](https://docs.databricks.com/en/sql/language-manual/data-types/special-floating-point-values.html).

For row-based window functions, these rules don't come into it, because the sorting is already complete.

For range-based window functions, the comparisons are made using the arithmetic rules listed.

Reply:
Thanks.